src/charm.py

```diff
@@ -683,539 +193,160 @@ def _report_ingress_ips(self) -> List[str]:
 
         Args:
-            current_svc_names: service names set by config or relation data.
-        """
-        api = self._core_v1_api()
-        all_services = api.list_namespaced_service(  # type: ignore[attr-defined]
-            namespace=self._namespace, label_selector=f"{CREATED_BY_LABEL}={self.app.name}"
-        )
-        all_svc_names = tuple(item.metadata.name for item in all_services.items)
-        unused_svc_names = tuple(
-            svc_name
-            for svc_name in all_svc_names
-            if svc_name.replace("-service", "") not in current_svc_names
-        )
-        LOGGER.debug(
-            "Checking for unused services. Configured: %s Found: %s Unused: %s",
-            current_svc_names,
-            all_svc_names,
-            unused_svc_names,
-        )
-        for unused_svc_name in unused_svc_names:
-            api.delete_namespaced_service(  # type: ignore[attr-defined]
-                name=unused_svc_name,
-                namespace=self._namespace,
-            )
-            LOGGER.info(
-                "Service deleted in namespace %s with name %s",
-                self._namespace,
-                unused_svc_name,
-            )
-
-    def _define_services(self) -> None:
-        """Create or update the services in Kubernetes from multiple ingress relations."""
-        for conf_or_rel in self._all_config_or_relations:
-            # By default, the service name is "". If there is no value set, we might be missing
-            # the needed relation data from that relation at this moment. Skip creating a Service
-            # for the current relation; it will be created when the relation data is set.
-            if self._has_required_fields(conf_or_rel):
-                self._define_service(conf_or_rel)
-
-    def _define_service(self, conf_or_rel: _ConfigOrRelation) -> None:
-        """Create or update a service in kubernetes.
+            controller: The controller for the resource type.
 
-        Args:
-            conf_or_rel: Ingress configuration or relation class.
+        Returns:
+            The name of the created or modified resource, None if no resource is
+            modified or created.
         """
-        api = self._core_v1_api()
-        body = conf_or_rel._get_k8s_service(self.app.name)
-        services = api.list_namespaced_service(  # type: ignore[attr-defined]
-            namespace=self._namespace
+        if self._ingress_options is None:
+            return None
+        resource_list = controller.list_resource(
+            namespace=self._namespace, label_selector=self._label_selector
         )
-        if conf_or_rel._k8s_service_name in [x.metadata.name for x in services.items]:
-            api.patch_namespaced_service(  # type: ignore[attr-defined]
-                name=conf_or_rel._k8s_service_name,
+        body = controller.gen_resource_from_options(self._ingress_options)
+        if body.metadata.name in [r.metadata.name for r in resource_list]:
```

Comment:
this shouldn't use a list it should use a set

Reply:
There will only be one or two resource in the list, the performance gain is not very compelling in this context.